### PR TITLE
Updated Augmentation API with random_seed

### PR DESCRIPTION
**Pull request by shijianjian**
The main updates are the user experiences for random augmentation by adding random_seed for each transformation for reproducibility. Behaviour changed from:
```python
trans = SOMETRANSFORM(**kwargs)
params = SOMETRANSFORM.get_params(param_range)
out_a = trans(input_a, params)
out_b = trans(input_b, params)
```
to:
```python
trans = SOMETRANSFORM(**kwargs, random_seed=seed)
out_a = trans(input_a)
out_b = trans(input_b)
```
or to override seed value
```python
trans = SOMETRANSFORM(**kwargs, random_seed=seed_1)
out_a = trans(input_a, random_seed=seed_2)
out_b = trans(input_b, random_seed=seed_2)
```

### CODE REVIEW

**Comment by edgarriba** [comment]
@shijianjian thanks for the PR.

Before reviewing, would like to share to you the feedback I got from @t-vi in pytorch slack channel  about the data augmentation API.

Screenshot_2020-02-22_23-47-56

Screenshot_2020-02-22_23-48-18

**Comment by shijianjian** [comment]
> @shijianjian thanks for the PR.
> 
> Before reviewing, would like to share to you the feedback I got from @t-vi in pytorch slack channel about the data augmentation API.
> 
> Screenshot_2020-02-22_23-47-56
> 
> Screenshot_2020-02-22_23-48-18

Indeed. I think he pointed out some issues I did not pay attention to. He seems complain about the accuracy of color jitter as well.

Anyway, in general, I have reviewed what we have:
- For torchvision mimicking:
    - Compose (N)
    - CenterCrop (Y)
    - ColorJitter(Y)
    - FiveCrop (N)
    - Grayscale (N)
    - Pad (N)
    - RandomAffine (Y&N, missing resample, fillcolor args)
    - RandomApply (N)
    - RandomChoice (N)
    - RandomCrop (Y)
    - RandomGrayscale (Y)
    - RandomHorizontalFlip (Y)
    - RandomOrder (N)
    - RandomPerspective(Y&N, missing interpolations)
    - RandomResizedCrop (Y&N, missing interpolations)
    - RandomRotation (Y&N, missing resamples)
    - RandomVerticalFlip (Y)
    - Resize (N)
    - TenCrop (N)
    - LinearTransformation (N)
    - Normalize (N)
    - RandomErasing (N,  can rename RandomRectangleEracsing to this)
- nn.Module-like classes were implemented in other files already. e.g. kornia.color.normalize.Normalize
- Calling modulelist inside ColorJitter is inappropriate.
- functional functions like ```adjust_hue``` were put in ```kornia.color.adjust.adjust_hue```, which is different from torchvision.
- Should we rewrite all the ```random*``` APIs then expose pure functional APIs as torchvision did?
- Input dtype. Are we accepting either 0-1 or uint8? If not appropriate, we can put the corresponding dtype conversions into ```kornia.augmentation.utils._transform_input()``` or add more functions in augmentation like ```FromUINT8Image()``` or ```FromFloatingImage()```.
- Unit testing may not be sufficient. We may need to compare the results with torchvision under the same parameter settings for accurate measurement.

**Comment by t-vi** [comment]
Hi,

so I didn't mean to make this put this on the record here unfiltered, it is more a description what I did to conclude that Kornia's augmentation wasn't the best fit for my project than anything else.

What I was looking for was more or less a 1-1 replacement for torchvision that
 - takes (batches of) `Tensor`s from TorchVision's ToTensor (data-loader collated),
 - replaces Transforms with nn.Modules,
 - uses PyTorch random functions exclusively for randomness (I really think that's a funny part about TorchVision),
 - works on GPU.

So my overall impression was that Kornia has quite a few more moving bits involved in the augmentation than TorchVision.
Some things were not handled (yet) well.
At least the contrast change seemed to be a brightness changes, which made me wonder about testing against torchvision.
I found the parameter handling funny, with assigning to self and all.

Personally, I think as long as you use PyTorch's RNG exclusively, there isn't a reason to feed a random seed to the augmentation itself. It is well-accepted that a global RNG state is a side-effect and it's probably enough to set a manual seed somewhere. If you want more finegrained control, taking an optional Generator instance instead looks like the most sane approach, look at the PyTorch API. A random seed is the wrong thing here!

But these are just my impressions of looking at the code for 30 minutes and wondering what is going on. You're much closer to Kornia and can tell what it needs.

**Comment by shijianjian** [comment]
Actually, as I integrated kornia augmentation to some of my projects, I found the most efficient way of doing it is to treat it as a component of my model than pre-processing augmentations.

The basic workflow comes with
1. Using torchvision for pre-processing. e.g. Resize, or so.
2. DataLoader collate to batches.
3. Feed to kornia augmentation module.
4. Feed to model.

I truly agree with you that we should add some tests against the torchvision output to ensure the same results can be obtained as identical. It seems like the results were messing around.

Also, as we are aiming for the reproducibility of augmentations as discussed [here](https://github.com/kornia/kornia/issues/312#issuecomment-570113696), which is a functionality that was not implemented in torchvision. You may refer to the [torch solution for the reproducibility](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html), which requires a lot of coding to do.

The current implementation included three methods to reproduce the transformations:
Firstly, using ```param_gen(pg)```:
```python
transform = SomeTransform(param_ranges)
params = pg.get_params(param_ranges)
ouput_a = transform(IMAGES_A, params)
ouput_b = transform(IMAGES_B, params)
```
Secondly, get the access to what have been done:
```python
transform = SomeTransform(param_ranges)
ouput_a = transform(IMAGES_A)
ouput_b = transform(IMAGES_B, transform._params)
```
Thirdly, using random_seed for each transform:
```python
transform = SomeTransform(param_ranges)
output_a = transform(IMAGES_A, random_seed=1)
output_b = transform(IMAGES_B, random_seed=1)
```
or make all transformations same:
```python
transform = SomeTransform(param_ranges, random_seed=1)
output_a = transform(IMAGES_A)
output_b = transform(IMAGES_B)
```

I have to admit that using random_seed is not always safe as I mentioned [here](https://github.com/kornia/kornia/pull/439#discussion_r378730655). It seems to be alright in the beginning since in most cases since the tensor is sequentially computed than multi-threads. After the augmentation part is done, the random_state will be restored. I will investigate if we can improve this behaviour by ```torch.Generator()```.

Thanks for your comments.

**Comment by edgarriba** [comment]
@t-vi I totally agree that the API and consequently its results should match with torchvision. In the very end we would like to integrate both APIs. I think that both frameworks could work altogether in this sense. However, I see kornia more the lib that implements all the functional stuff, that later frameworks like torchvision, fastai (or others) can consume in a more high-level abstraction. In fact, I see the whole data augmentation module in kornia already a high level (even too high level for kornia :) factory that generates the random parameters to later call whatever low level feature in kornia to perform the operators. This is my overview of the project :)

**Comment by edgarriba** [comment]
@shijianjian I'll answer by points your points.

> nn.Module-like classes were implemented in other files already. e.g. kornia.color.normalize.Normalize

As mention above, the data augmentation module in theory should just generate parameters, but functionals is whole kornia :) One thing we discussed at some points is that maybe we should have a `kornia.nn` module that exposes the functionals to nn.Module instead of mixing everything in every single module. 

> Calling modulelist inside ColorJitter is inappropriate.

totally agree, feel free to call functional instead. Besides, what I find more important is that all the internal adjust_* is well tested. I don't see ColorJitter class appropriated at all, because in the end it's just a combination of 4 out of all the different color transformation we have. It's nice to have it, but zero flexible in case you want to apply or combine more stuff. I'm in for deprecating it at the very end and having more smart compose thing (like fastai did) What are your views guys ? /cc @fmassa @jph00

> functional functions like adjust_hue were put in kornia.color.adjust.adjust_hue, which is different from torchvision.

from the point of view of a data augmentation library like torchvision, albumentations, [rising](https://github.com/PhoenixDL/rising) (check this last, got new guys in the room) adjust* stuff is a color transformation (I'd split in color/geometric),. However, as computer vision library I feel it's in the correct place since it has to do with pixel intensity changes and not geometry. Others like, [scikit-image](https://scikit-image.org/docs/stable/api/skimage.exposure.html) put this stuff in `exposure`. not sure to me it's a good name, but I'm open to create a similar module, or submodule under color.

> Should we rewrite all the random* APIs then expose pure functional APIs as torchvision 

for reproducibility probably yes 

> Input dtype. Are we accepting either 0-1 or uint8? If not appropriate, we can put the corresponding dtype conversions into kornia.augmentation.utils._transform_input() or add more functions in augmentation like FromUINT8Image() or FromFloatingImage()

this is a design decision we took with with @ducha-aiki at the very beginning. Kornia is meant for training nets, so no need of uint8. Apart from that, some of the torch operator require floating point precision. Maybe things are drifting and should rethink. Open to discuss /cc @ducha-aiki 

> Unit testing may not be sufficient. We may need to compare the results with torchvision under the same parameter settings for accurate measurement.

absolutely yes, same as we started doing with color conversions with opencv

**Comment by jph00** [comment]
> totally agree, feel free to call functional instead. Besides, what I find more important is that all the internal adjust_* is well tested. I don't see ColorJitter class appropriated at all, because in the end it's just a combination of 4 out of all the different color transformation we have. It's nice to have it, but zero flexible in case you want to apply or combine more stuff. I'm in for deprecating it at the very end and having more smart compose thing (like fastai did) What are your views guys ? /cc @fmassa <https://github.com/fmassa> @jph00 <https://github.com/jph00>


fastcore is a fairly minimal dependency if you want to use our composition pipeline. It makes a lot of things much easier.

**Comment by edgarriba** [comment]
@jph00 thanks ! will take a look at it. However, I think that including dependencies is out of the scope. We want to keep it clean, just pytorch and implement very low level things. In fact, this data augmentation module is even to high level to me, at least for the initial purpose of the librayr. Has this api or any other operators (existing or future) any interest to fastai (as consumers) ?

**Comment by ducha-aiki** [comment]
@edgarriba  fastcore is pytorch-only library as well. On the other hand, fast* libraries are changing faster than kornia, so it could be hard to maintain given our resources now.  Is fastcore an exception?  cc: @jph00 

Regarding uint8, I think it is reasonable to not support them, given that lots of pytorch-GPU functions do not support them either.

>However, I see kornia more the lib that implements all the functional stuff, that later frameworks like torchvision, fastai (or others) can consume in a more high-level abstraction. In fact, I see the whole data augmentation module in kornia already a high level (even too high level for kornia

Yes. There are different design choices for the high-level operations and we would like to allow all of them by not making the choice, allowing next-level libraries to do it instead. 

>Personally, I think as long as you use PyTorch's RNG exclusively, there isn't a reason to feed a random seed to the augmentation itself. 

 I agree

**Comment by jph00** [comment]
On Sun, Feb 23, 2020, at 1:11 PM, Dmytro Mishkin wrote:
> @edgarriba <https://github.com/edgarriba> fastcore is pytorch-only library as well. On the other hand, fast* libraries are changing faster than kornia, so it could be hard to maintain given our resources now. Is fastcore an exception? cc: @jph00 <https://github.com/jph00>


It's pretty stable now. We've had one significant change to Transform, removing the `as_item` parameter (which effected very few people in practice), but other than that nothing much has changed: https://github.com/fastai/fastcore/commits/master/fastcore/transform.py

**Comment by shijianjian** [comment]
> However, I see kornia more the lib that implements all the functional stuff, that later frameworks like torchvision, fastai (or others) can consume in a more high-level abstraction. In fact, I see the whole data augmentation module in kornia already a high level (even too high level for kornia :) factory that generates the random parameters to later call whatever low level feature in kornia to perform the operators. This is my overview of the project :)

To be honest, if the aim of Kornia is to have a GPU version of OpenCV then the whole augmentation module is not quite reasonable to exist. To have something like kornia.nn or another repo sounds better. Actually, I thought what we were doing here is more close to the [rising](https://github.com/PhoenixDL/rising) library you mentioned. It is better to make a decision that if we are going to give a torchvision-like API for data augmentation or not :)


> > Personally, I think as long as you use PyTorch's RNG exclusively, there isn't a reason to feed a random seed to the augmentation itself.
> 
> I agree

I have re-implemented a local version for RNG state. The whole pipeline comes like:
```python
if random_seed is provided:
    rng = torch.Generator()
    rng.manual_seed(random_seed)
    params = param_gen(rng, PARAM_SETTINGS)   # in local rng setting.
else:
    parmas = param_gen(PARAM_SETTINGS)   # in global rng setting.
```
This will guarentee the reproducibility if we provide a random_seed without affecting the global state. The purpose of this control is the user can easily train a model if he/she needs the same augmentation, e.g. in semantic segmentation, etc.

**Comment by edgarriba** [comment]
@shijianjian should we enable to pass random seed to top layer to reproduce ? othwerwise close this PR

**Comment by shijianjian** [comment]
For reproducing, I prefer to feed the same params directly than having an additional ```seed``` parameter. I will close this.
